networkx/utils/backends.py

```diff
@@ -972,20 +976,162 @@ def _convert_arguments(self, backend_name, args, kwargs):
+        backend_name,
+        graph,
+        *,
+        edge_attrs,
+        node_attrs,
+        preserve_edge_attrs,
+        preserve_node_attrs,
+        preserve_graph_attrs,
+        graph_name,
+        use_cache,
+    ):
+        if (
+            use_cache
+            and (nx_cache := getattr(graph, "__networkx_cache__", None)) is not None
+        ):
+            cache = nx_cache.setdefault("backends", {}).setdefault(backend_name, {})
+            # edge_attrs: dict | None
+            # node_attrs: dict | None
+            # preserve_edge_attrs: bool (False if edge_attrs is not None)
+            # preserve_node_attrs: bool (False if node_attrs is not None)
+            # preserve_graph_attrs: bool
+            key = edge_key, node_key, graph_key = (
+                frozenset(edge_attrs.items())
+                if edge_attrs is not None
+                else preserve_edge_attrs,
+                frozenset(node_attrs.items())
+                if node_attrs is not None
+                else preserve_node_attrs,
+                preserve_graph_attrs,
+            )
+            if cache:
+                # Do a simple search for a cached graph with compatible data.
+                # For example, if we need a single attribute, then it's okay
+                # to use a cached graph that preserved all attributes.
+                # This looks for an exact match first.
+                for compat_key in itertools.product(
+                    (edge_key, True) if edge_key is not True else (True,),
+                    (node_key, True) if node_key is not True else (True,),
+                    (graph_key, True) if graph_key is not True else (True,),
+                ):
+                    if (rv := cache.get(compat_key)) is not None:
+                        warnings.warn(
+                            f"Using cached graph for {backend_name!r} backend in "
+                            f"call to {self.name}.\n\nFor the cache to be consistent "
+                            "(i.e., correct), the input graph must not have been "
+                            "manually mutated since the cached graph was created. "
+                            "Examples of manually mutating the graph data structures "
+                            "resulting in an inconsistent cache include:\n\n"
+                            "    >>> G[u][v][key] = val\n\n"
+                            "and\n\n"
+                            "    >>> for u, v, d in G.edges(data=True):\n"
+                            "    ...     d[key] = val\n\n"
+                            "Using methods such as `G.add_edge(u, v, weight=val)` "
+                            "will correctly clear the cache to keep it consistent. "
+                            "You may also use `G.__networkx_cache__.clear()` to "
+                            "manually clear the cache, or set `G.__networkx_cache__` "
+                            "to None to disable caching for G. Enable or disable "
```

Comment:
Dan suggested this here: https://github.com/networkx/networkx/pull/7345#issuecomment-2008562494

Would it be better to suggest `nx._clear_cache(G)` instead?

Reply:
I do like that better, but I don't feel strongly either way.  I'm also thinking we can keep it named with the sunder since it's easier to go private->public later.